src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,297 @@
+            /// This refactoring only supports non-linked document and linked document in the form of
+            /// documents in multi-targeting project. Also for simplicity, we also don't support document
+            /// what has different file path and logical path in project (i.e. [ProjectRoot] + `Document.Folders`). 
+            /// If the requirements above is met, we will return IDs of all documents linked to the specified 
+            /// document (inclusive), an array of single element will be returned for non-linked document.
+            /// </summary>
+            private static bool IsSupportedLinkedDocument(Document document, out ImmutableArray<DocumentId> allDocumentIds)
+            {
+                var solution = document.Project.Solution;
+                var linkedDocumentids = document.GetLinkedDocumentIds();
+
+                // TODO: figure out how to properly determine if a document is linked using project system.
+
+                // If we found a linked document which is part of a project with differenct project file,
+                // then it's an actual linked file (i.e. not a MTFM project). We don't support that, because 
+                // we don't know which default namespace and folder path we should use to construct target
+                // namespace.
+                if (linkedDocumentids.Any(id => 
+                !PathUtilities.PathsEqual(solution.GetDocument(id).Project.FilePath, document.Project.FilePath)))
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+
+                // Now determine if the actual file path matches its logical path in project 
+                // which is constructed as <project root path>\Logical\Folders\. The refactoring 
+                // is triggered only when the two match. The reason of doing this is we don't really know
+                // the user's intention of keeping the file path out-of-sync with its logical path.
+                var projectRoot = PathUtilities.GetDirectoryName(document.Project.FilePath);
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                if (PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath))
+                {
+                    allDocumentIds = linkedDocumentids.Add(document.Id);
+                    return true;
+                }
+                else
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+            }
+
+            private async static Task<(bool shouldTriggerRefactoring, string defaultNamespace, string declaredNamespace)> 
+                TryGetDefaultNamespaceAndNamespaceDeclarationAsync(
```

Comment:
this is super confusing as this looks like a parameter.  If we have such complex data to return, consdier just having a real type here.